cardano-node/src/Cardano/Node/Protocol/Checkpoints.hs

```diff
@@ -0,0 +1,115 @@
+import           Cardano.Node.Types
+import           Ouroboros.Consensus.Block
+import           Ouroboros.Consensus.Cardano
+import           Ouroboros.Consensus.Config (CheckpointsMap (..), emptyCheckpointsMap)
+
+import           Control.Exception (IOException)
+import           Control.Monad (forM, unless, when)
+import qualified Data.Aeson as Aeson
+import qualified Data.Aeson.Types as Aeson
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Base16 as B16
+import           Data.Foldable (forM_)
+import qualified Data.Map.Strict as Map
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import qualified Data.Text as Text
+import qualified Data.Text.Encoding as Text
+
+data CheckpointsReadError =
+       CheckpointsReadFileError !FilePath !IOException
+     | CheckpointsHashMismatch
+         !FilePath
+         -- | Actual
+         !CheckpointsHash
+         -- | Expected
+         !CheckpointsHash
+     | CheckpointsDecodeError !FilePath !String
+   deriving Show
+
+instance Error CheckpointsReadError where
+  prettyError (CheckpointsReadFileError fp err) =
+        "There was an error reading the checkpoints file: "
+     <> pshow fp <> " Error: " <> pshow err
+
+  prettyError (CheckpointsHashMismatch fp actual expected) =
+        "Hash mismatch for checkpoints file " <> pshow fp <> ": "
+     <> "the actual hash is " <> pshow actual <> ", but the expected "
+     <> "hash given in the node configuration file is " <> pshow expected
+
+  prettyError (CheckpointsDecodeError fp err) =
+        "There was an error parsing the checkpoints file: "
+     <> pshow fp <> " Error: " <> pshow err
+
+readCheckpointsMap
+  :: NodeCheckpointsConfiguration
+  -> ExceptT CheckpointsReadError IO (CheckpointsMap (CardanoBlock StandardCrypto))
+readCheckpointsMap NodeCheckpointsConfiguration {
+                       npcCheckpointsFile,
+                       npcCheckpointsFileHash = mExpectedHash
+                     } = case npcCheckpointsFile of
+    Nothing -> pure emptyCheckpointsMap
+    Just (CheckpointsFile file) -> do
+      content <-
+        handleIOExceptT (CheckpointsReadFileError file) $ BS.readFile file
+
+      let actualHash = CheckpointsHash $ Crypto.hashWith id content
+      forM_ mExpectedHash $ \expectedHash ->
+        when (actualHash /= expectedHash) $
+          throwError (CheckpointsHashMismatch file actualHash expectedHash)
+
+      WrapCheckpointsMap checkpointsMap <-
+        firstExceptT (CheckpointsDecodeError file) $ hoistEither $
+          Aeson.eitherDecodeStrict' content
+      pure checkpointsMap
+
+newtype WrapCheckpointsMap =
+        WrapCheckpointsMap (CheckpointsMap (CardanoBlock StandardCrypto))
+
+instance Aeson.FromJSON WrapCheckpointsMap where
+  parseJSON = Aeson.withObject "CheckpointsMap" $ \o -> do
+      checkpointList :: [Aeson.Object] <- o Aeson..: "checkpoints"
+
+      checkpoints :: [(BlockNo, HeaderHash (CardanoBlock StandardCrypto))] <-
```

Comment:
Personally I'd consider a `{blockNoHashPairs:  [[3, "abcdef"], [100, "32cdef"]]}` format, to avoid the obnoxious per-entry repetition of the field names. But I don't have a lot of experience regarding the widely-expected UX for this sort of thing.

Reply:
Yeah :+1:, I have no strong opinions here; one tiny advantage of the current format is that it is impossible to forgot that the numbers are referring to block numbers (and not e.g. slot numbers), even when looking at individual checkpoints only; but that is not a major thing.